src/core/src/lib.rs

```diff
@@ -336,25 +405,52 @@ pub trait QueueFamily: 'static {
     fn num_queues(&self) -> u32;
 }
 
+/// Submission information for a command queue.
+pub struct QueueSubmit<'a, C: CommandBuffer + 'a, R: Resources> {
+    /// Command buffers to submit.
+    pub cmd_buffers: &'a [command::Submit<C>],
+    /// Semaphores to wait being signaled before submission.
+    pub wait_semaphores: &'a [(&'a mut R::Semaphore, pso::PipelineStage)],
+    /// Semaphores which get signaled after submission.
+    pub signal_semaphores: &'a [&'a mut R::Semaphore],
+}
+
 /// Dummy trait for command queues.
 /// CommandBuffers will be later submitted to command queues instead of the device.
-pub trait CommandQueue { }
+pub trait CommandQueue {
+    /// Associated `Resources` type.
+    type Resources: Resources;
+    /// Submit data for command queues.
+    type SubmitInfo;
+    /// General command buffer type of the backend.
+    type GeneralCommandBuffer: CommandBuffer<SubmitInfo = Self::SubmitInfo>; // + GraphicsCommandBuffer<Self::R> + ComputeCommandBuffer<Self::R>;
+    /// Graphics command buffer type of the backend.
+    type GraphicsCommandBuffer: CommandBuffer<SubmitInfo = Self::SubmitInfo>; // + GraphicsCommandBuffer<Self::R>;
+    /// Compute command buffer type of the backend.
+    type ComputeCommandBuffer: CommandBuffer<SubmitInfo = Self::SubmitInfo>; // + ComputeCommandBuffer<Self::R>;
+    /// Transfer/Copy command buffer type of the backend.
+    type TransferCommandBuffer: CommandBuffer<SubmitInfo = Self::SubmitInfo>; // + TransferCommandBuffer<Self::R>;
+    /// Subpass command buffer type of the backend.
+    type SubpassCommandBuffer: CommandBuffer<SubmitInfo = Self::SubmitInfo>; // + SubpassCommandBuffer<Self::R>;
+
+    /// Submit command buffers to queue for execution.
+    unsafe fn submit<'a, C>(&mut self, submit_infos: &[QueueSubmit<C, Self::Resources>], fence: Option<&'a mut <Self::Resources as Resources>::Fence>)
+        where C: CommandBuffer<SubmitInfo = Self::SubmitInfo>;
+}
 
 /// A `Surface` abstracts the surface of a native window, which will be presented
 pub trait Surface {
     /// Associated `CommandQueue` type.
     type CommandQueue: CommandQueue;
-    /// Associated `SwapChain` type.
-    type SwapChain: SwapChain;
     /// Associated native `Window` type.
     type Window;
 
     /// Create a new surface from a native window.
     fn from_window(window: &Self::Window) -> Self;
 
     /// Create a new swapchain from the current surface with an associated present queue.
-    fn build_swapchain<T: format::RenderFormat>(&self, present_queue: &Self::CommandQueue)
-        -> Self::SwapChain;
+    fn build_swapchain<S, T>(&self, present_queue: &Self::CommandQueue) -> S
+        where S: SwapChain, T: format::RenderFormat;
```

Comment:
Surface got some further changes when I implemented the GL and vulkan backend.
The current Swapchain generic got removed again.
Surfaces will be moved towards the current window crates and one of the two entry points for creation (besides Adapter).
So the user is supposed to create a surface from a window (usually winit::Window) and further create an associated swapchain. The swapchain has to be recreated on resize or other events.